File: source/Canon50M/shutter.c
#include <linux/ioctl.h>
#include <sys/ioctl.h>
#include <unistd.h>
#include <stdio.h>
#include <errno.h> 
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdlib.h>

#include <getopt.h>

#define GP0_OUTPUT_ENABLE	_IOW(0x81,0,int)
#define GP0_SET_VALUE		_IOW(0x81,1,int)
#define GP0_GET_VALUE		_IOR(0x81,2,int)

#define S_CLOSE  0
#define S_OPEN   1

/* Print out the proper program usage syntax */
static void
printUsageSyntax(char *prgname) {
   fprintf(stderr,
	   "Metrology camera shutter control.\n"
	   "Usage: %s [options...]\n"
       "	-o, --open     Open shutter.\n"
       "	-c, --close    Close shutter.\n"
       "	-t, --time     Set timimg for shutter (ms).\n"
	   "	-h, --help     display help message.\n"
	   "	-v, --verbose  turn on verbose.\n"
		, prgname);

}


int main(int argc, char* argv[]){
	int opt;
	int i,verbose=0;
	int shutter=S_CLOSE;
	int time=0;

	int iOUTPUT_ENABLE=1;
	int outputData;
	int res;

	/** Check the total number of the arguments */
	struct option longopts[] = {
		 {"open" ,0, NULL, 'o'},
		 {"close" ,0, NULL, 'c'},
		 {"time" ,1, NULL, 't'},
		 {"verbose",0, NULL, 'v'},
		 {"help", 0, NULL, 'h'},
		 {0,0,0,0}};

	while((opt = getopt_long(argc, argv, "oct:vh",
	   longopts, NULL))  != -1){
	      switch(opt) {
	         case 'o':
	               shutter = S_OPEN;
	               break;
	         case 'c':
	               shutter = S_CLOSE;
	               break;
	         case 't':
	               time = atoi(optarg);
	               shutter = -1;
	               break;
	         case 'v':
	               verbose = 1;
	               break;
	         case 'h':
	               printUsageSyntax(argv[0]);
	               exit(EXIT_FAILURE);
	               break;
	         case '?':
	               printUsageSyntax(argv[0]);
	               exit(EXIT_FAILURE);
	               break;
	      }
	}


	/** Print the usage syntax if there is no input */
	if (argc < 2 ) {
		printUsageSyntax(argv[0]);
		return EXIT_FAILURE;
	}

	if (time !=0 && time < 10 ) {
		printf("Number too small. The unit of time is ms!!!\n");
		return EXIT_FAILURE;
	}

	i = open("/dev/ttyUSB0", O_RDWR);
	if (i==-1) {
		printf("Cannot open device!!!\n");
		return -1;
	}

	/* Enable GPIO_1 to output mode*/
	res = ioctl(i, GP0_OUTPUT_ENABLE,&iOUTPUT_ENABLE);
	if (res==-1) {
		printf("Unable to access shutter, check device.\n");
		return EXIT_FAILURE;	
	}

	if (shutter == S_OPEN){
		printf("Opening shutter.\n");
		outputData=S_OPEN;
		res = ioctl(i, GP0_SET_VALUE,&outputData);
	}

	if (shutter == S_CLOSE){
		printf("closing shutter.\n");
		outputData=S_CLOSE;
		res = ioctl(i, GP0_SET_VALUE,&outputData);

	}

	if (time != 0){
		printf("Opening shutter for %d msec.\n",time);
		outputData=1;
		res = ioctl(i, GP0_SET_VALUE,&outputData);
		usleep(time*1000);
		outputData=0;

		res = ioctl(i, GP0_SET_VALUE,&outputData);

	}

	close(i);
	return 0;
}

File: source/Canon50M/shutter.py
import sys
import getopt
import time
import fcntl
import struct
import os

# Updated command numbers based on the C program output
GP0_OUTPUT_ENABLE = 0x40048100
GP0_SET_VALUE = 0x40048101
GP0_GET_VALUE = 0x80048102
S_CLOSE = 0
S_OPEN = 1

def print_usage_syntax(program_name):
    print(f"""Metrology camera shutter control.
Usage: {program_name} [options...]
    -o, --open     Open shutter.
    -c, --close    Close shutter.
    -t, --time     Set timing for shutter (ms).
    -d, --device   Specify device path (default: /dev/ttyUSB0).
    -h, --help     Display help message.
    -v, --verbose  Turn on verbose.""", file=sys.stderr)

def ioctl_command(fd, command, value):
    try:
        result = fcntl.ioctl(fd, command, struct.pack('I', value))
        return True
    except IOError as e:
        print(f"ioctl failed. Command: {hex(command)}, Value: {value}, Error: {e}")
        return False

def main(argv):
    verbose = False
    shutter = S_CLOSE
    timing = 0
    device_path = "/dev/ttyUSB0"

    try:
        opts, args = getopt.getopt(argv[1:], "oct:d:vh", ["open", "close", "time=", "device=", "verbose", "help"])
    except getopt.GetoptError:
        print_usage_syntax(argv[0])
        sys.exit(2)

    for opt, arg in opts:
        if opt in ("-o", "--open"):
            shutter = S_OPEN
        elif opt in ("-c", "--close"):
            shutter = S_CLOSE
        elif opt in ("-t", "--time"):
            timing = int(arg)
            shutter = -1
        elif opt in ("-d", "--device"):
            device_path = arg
        elif opt in ("-v", "--verbose"):
            verbose = True
        elif opt in ("-h", "--help"):
            print_usage_syntax(argv[0])
            sys.exit()

    if len(argv) < 2:
        print_usage_syntax(argv[0])
        sys.exit(2)

    if timing != 0 and timing < 10:
        print("Number too small. The unit of time is ms!!!")
        sys.exit(2)

    try:
        fd = os.open(device_path, os.O_RDWR)
        print(f"Successfully opened device: {device_path}")

        # Enable GPIO_1 to output mode
        if not ioctl_command(fd, GP0_OUTPUT_ENABLE, 1):
            raise IOError("Failed to set GPIO to output mode")

        if verbose:
            print(f"GP0_OUTPUT_ENABLE = {GP0_OUTPUT_ENABLE} {hex(GP0_OUTPUT_ENABLE)}")
            print(f"GP0_SET_VALUE = {GP0_SET_VALUE} {hex(GP0_SET_VALUE)}")
            print(f"GP0_GET_VALUE = {GP0_GET_VALUE} {hex(GP0_GET_VALUE)}")

        if shutter == S_OPEN:
            print("Opening shutter.")
            if not ioctl_command(fd, GP0_SET_VALUE, S_OPEN):
                raise IOError("Failed to open shutter")
        elif shutter == S_CLOSE:
            print("Closing shutter.")
            if not ioctl_command(fd, GP0_SET_VALUE, S_CLOSE):
                raise IOError("Failed to close shutter")
        elif timing != 0:
            print(f"Opening shutter for {timing} msec.")
            if not ioctl_command(fd, GP0_SET_VALUE, S_OPEN):
                raise IOError("Failed to open shutter")
            time.sleep(timing / 1000)
            if not ioctl_command(fd, GP0_SET_VALUE, S_CLOSE):
                raise IOError("Failed to close shutter")

    except Exception as e:
        print(f"An error occurred: {e}")
        sys.exit(1)
    finally:
        os.close(fd)
        print(f"Closed device: {device_path}")

if __name__ == "__main__":
    main(sys.argv)